File: python/benchmarks/benchmark_binary_heap.py
import heapq
import random
import time
from typing import List, Tuple

from binary_heap import BinaryHeap
from common import make_stdout_csv_writer, write_csv_header, benchmark_case, make_sequential_values, make_reverse_sequential_values, make_random_unique_values, make_shuffled


# -----------------------------
# Общие настройки
# -----------------------------

SIZES = [10_000, 50_000, 100_000, 200_000, 500_000, 1_000_000, 2_000_000, 5_000_000, 10_000_000, 15_000_000]
REPEATS = 5
RANDOM_SEED = 42
TOP_K = 100


# -----------------------------
# 1. Push sequential
# -----------------------------

def prepare_case_push_sequential(n: int) -> Tuple[List[int], BinaryHeap, List[int]]:
    values = make_sequential_values(n)
    return values, BinaryHeap(), []


def measure_push_sequential_binary_heap(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, base_heap, _ = prepared
    heap = base_heap.copy()

    start = time.perf_counter()
    for value in values:
        heap.push(value)
    end = time.perf_counter()

    return end - start


def measure_push_sequential_heapq(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, _, base_heapq = prepared
    heap = base_heapq.copy()

    start = time.perf_counter()
    for value in values:
        heapq.heappush(heap, value)
    end = time.perf_counter()

    return end - start


# -----------------------------
# 2. Push reverse sequential
# -----------------------------

def prepare_case_push_reverse_sequential(n: int) -> Tuple[List[int], BinaryHeap, List[int]]:
    values = make_reverse_sequential_values(n)
    return values, BinaryHeap(), []


def measure_push_reverse_sequential_binary_heap(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, base_heap, _ = prepared
    heap = base_heap.copy()

    start = time.perf_counter()
    for value in values:
        heap.push(value)
    end = time.perf_counter()

    return end - start


def measure_push_reverse_sequential_heapq(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, _, base_heapq = prepared
    heap = base_heapq.copy()

    start = time.perf_counter()
    for value in values:
        heapq.heappush(heap, value)
    end = time.perf_counter()

    return end - start


# -----------------------------
# 3. Push random
# -----------------------------

def prepare_case_push_random(n: int) -> Tuple[List[int], BinaryHeap, List[int]]:
    values = make_random_unique_values(n, RANDOM_SEED)
    return values, BinaryHeap(), []


def measure_push_random_binary_heap(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, base_heap, _ = prepared
    heap = base_heap.copy()

    start = time.perf_counter()
    for value in values:
        heap.push(value)
    end = time.perf_counter()

    return end - start


def measure_push_random_heapq(prepared: Tuple[List[int], BinaryHeap, List[int]]) -> float:
    values, _, base_heapq = prepared
    heap = base_heapq.copy()

    start = time.perf_counter()
    for value in values:
        heapq.heappush(heap, value)
    end = time.perf_counter()

    return end - start


# -----------------------------
# 4. Heapify random
# -----------------------------

def prepare_case_heapify_random(n: int) -> List[int]:
    return make_random_unique_values(n, RANDOM_SEED)


def measure_heapify_random_binary_heap(values: List[int]) -> float:
    data = values.copy()

    start = time.perf_counter()
    _ = BinaryHeap(data)
    end = time.perf_counter()

    return end - start


def measure_heapify_random_heapq(values: List[int]) -> float:
    data = values.copy()

    start = time.perf_counter()
    heapq.heapify(data)
    end = time.perf_counter()

    return end - start


# -----------------------------
# 5. Pop all after heapify
# -----------------------------

def prepare_case_pop_all_after_heapify(n: int) -> Tuple[BinaryHeap, List[int]]:
    values = make_random_unique_values(n, RANDOM_SEED)

    base_heap = BinaryHeap(values)

    base_heapq = values.copy()
    heapq.heapify(base_heapq)

    return base_heap, base_heapq


def measure_pop_all_after_heapify_binary_heap(prepared: Tuple[BinaryHeap, List[int]]) -> float:
    base_heap, _ = prepared
    heap = base_heap.copy()

    start = time.perf_counter()
    while len(heap) > 0:
        heap.pop()
    end = time.perf_counter()

    return end - start


def measure_pop_all_after_heapify_heapq(prepared: Tuple[BinaryHeap, List[int]]) -> float:
    _, base_heapq = prepared
    heap = base_heapq.copy()

    start = time.perf_counter()
    while len(heap) > 0:
        heapq.heappop(heap)
    end = time.perf_counter()

    return end - start


# -----------------------------
# 6. Alternating push/pop
# -----------------------------

def prepare_case_alternating_push_pop(n: int) -> Tuple[List[Tuple[str, int]], BinaryHeap, List[int]]:
    rng = random.Random(RANDOM_SEED)

    initial_values = make_random_unique_values(n // 2, RANDOM_SEED)
    operation_values = make_random_unique_values(n, RANDOM_SEED + 1)

    operations = []
    current_size = len(initial_values)

    for value in operation_values:
        if current_size == 0:
            operations.append(("push", value))
            current_size += 1
            continue

        if rng.random() < 0.5:
            operations.append(("push", value))
            current_size += 1
        else:
            operations.append(("pop", value))
            current_size -= 1

    base_heap = BinaryHeap(initial_values)

    base_heapq = initial_values.copy()
    heapq.heapify(base_heapq)

    return operations, base_heap, base_heapq


def measure_alternating_push_pop_binary_heap(prepared: Tuple[List[Tuple[str, int]], BinaryHeap, List[int]]) -> float:
    operations, base_heap, _ = prepared
    heap = base_heap.copy()

    start = time.perf_counter()

    for operation, value in operations:
        if operation == "push":
            heap.push(value)
        else:
            heap.pop()

    end = time.perf_counter()

    return end - start


def measure_alternating_push_pop_heapq(prepared: Tuple[List[Tuple[str, int]], BinaryHeap, List[int]]) -> float:
    operations, _, base_heapq = prepared
    heap = base_heapq.copy()

    start = time.perf_counter()

    for operation, value in operations:
        if operation == "push":
            heapq.heappush(heap, value)
        else:
            heapq.heappop(heap)

    end = time.perf_counter()

    return end - start


# -----------------------------
# 7. Top-k largest
# -----------------------------

def prepare_case_top_k_largest(n: int) -> Tuple[List[int], int]:
    values = make_random_unique_values(n, RANDOM_SEED)
    k = min(TOP_K, n)
    return values, k


def measure_top_k_largest_binary_heap(prepared: Tuple[List[int], int]) -> float:
    values, k = prepared
    heap = BinaryHeap()

    start = time.perf_counter()

    for value in values:
        if len(heap) < k:
            heap.push(value)
        elif value > heap.peek():
            heap.pop()
            heap.push(value)

    end = time.perf_counter()

    return end - start


def measure_top_k_largest_heapq(prepared: Tuple[List[int], int]) -> float:
    values, k = prepared
    heap = []

    start = time.perf_counter()

    for value in values:
        if len(heap) < k:
            heapq.heappush(heap, value)
        elif value > heap[0]:
            heapq.heappop(heap)
            heapq.heappush(heap, value)

    end = time.perf_counter()

    return end - start


# -----------------------------
# Запуск всех benchmark-ов
# -----------------------------

def run_all() -> None:
    writer = make_stdout_csv_writer()
    write_csv_header(writer)

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="1. Push sequential",
        sizes=SIZES,
        prepare_case=prepare_case_push_sequential,
        implementations=[("BinaryHeap", measure_push_sequential_binary_heap), ("heapq", measure_push_sequential_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="2. Push reverse sequential",
        sizes=SIZES,
        prepare_case=prepare_case_push_reverse_sequential,
        implementations=[("BinaryHeap", measure_push_reverse_sequential_binary_heap), ("heapq", measure_push_reverse_sequential_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="3. Push random",
        sizes=SIZES,
        prepare_case=prepare_case_push_random,
        implementations=[("BinaryHeap", measure_push_random_binary_heap), ("heapq", measure_push_random_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="4. Heapify random",
        sizes=SIZES,
        prepare_case=prepare_case_heapify_random,
        implementations=[("BinaryHeap", measure_heapify_random_binary_heap), ("heapq", measure_heapify_random_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="5. Pop all after heapify",
        sizes=SIZES,
        prepare_case=prepare_case_pop_all_after_heapify,
        implementations=[("BinaryHeap", measure_pop_all_after_heapify_binary_heap), ("heapq", measure_pop_all_after_heapify_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="6. Alternating push/pop",
        sizes=SIZES,
        prepare_case=prepare_case_alternating_push_pop,
        implementations=[("BinaryHeap", measure_alternating_push_pop_binary_heap), ("heapq", measure_alternating_push_pop_heapq)],
        repeats=REPEATS,
    )

    benchmark_case(
        writer=writer,
        structure="binary_heap",
        case_name="7. Top-k largest",
        sizes=SIZES,
        prepare_case=prepare_case_top_k_largest,
        implementations=[("BinaryHeap", measure_top_k_largest_binary_heap), ("heapq", measure_top_k_largest_heapq)],
        repeats=REPEATS,
    )


if __name__ == "__main__":
    run_all()

File: cpp/benchmarks/benchmark_binary_heap.cpp
#include <chrono>
#include <functional>
#include <queue>
#include <random>
#include <string>
#include <utility>
#include <vector>

#include "common.hpp"
#include "binary_heap.h"

using namespace std;

using Time = chrono::high_resolution_clock;
using Seconds = chrono::duration<double>;
using MinPriorityQueue = priority_queue<int, vector<int>, greater<int>>;

// This variable prevents the optimizer from removing benchmarked work
// when the result of operations is not otherwise used.
static volatile size_t benchmark_sink = 0;

// -----------------------------
// Общие настройки
// -----------------------------

const vector<int> SIZES = {10'000, 50'000, 100'000, 200'000, 500'000, 1'000'000, 2'000'000, 5'000'000, 10'000'000, 15'000'000};
const int REPEATS = 5;
const int RANDOM_SEED = 42;
const int TOP_K = 100;

// -----------------------------
// Вспомогательные структуры
// -----------------------------

struct BasicPrepared {
    vector<int> values;
    BinaryHeap<int> heap;
    MinPriorityQueue std_heap;
};

struct HeapifyPrepared {
    vector<int> values;
};

struct PopAllPrepared {
    BinaryHeap<int> heap;
    MinPriorityQueue std_heap;
};

struct AlternatingPrepared {
    vector<pair<string, int>> operations;
    BinaryHeap<int> heap;
    MinPriorityQueue std_heap;
};

struct TopKPrepared {
    vector<int> values;
    int k;
};

// -----------------------------
// 1. Push sequential
// -----------------------------

BasicPrepared prepare_case_push_sequential(int n) {
    return {make_sequential_values(n), BinaryHeap<int>(), MinPriorityQueue()};
}

double measure_push_sequential_binary_heap(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    BinaryHeap<int> heap = prepared.heap.copy();

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

double measure_push_sequential_std_priority_queue(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    MinPriorityQueue heap = prepared.std_heap;

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// 2. Push reverse sequential
// -----------------------------

BasicPrepared prepare_case_push_reverse_sequential(int n) {
    return {make_reverse_sequential_values(n), BinaryHeap<int>(), MinPriorityQueue()};
}

double measure_push_reverse_sequential_binary_heap(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    BinaryHeap<int> heap = prepared.heap.copy();

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

double measure_push_reverse_sequential_std_priority_queue(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    MinPriorityQueue heap = prepared.std_heap;

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// 3. Push random
// -----------------------------

BasicPrepared prepare_case_push_random(int n) {
    return {make_random_unique_values(n, RANDOM_SEED), BinaryHeap<int>(), MinPriorityQueue()};
}

double measure_push_random_binary_heap(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    BinaryHeap<int> heap = prepared.heap.copy();

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

double measure_push_random_std_priority_queue(const BasicPrepared& prepared) {
    const vector<int>& values = prepared.values;
    MinPriorityQueue heap = prepared.std_heap;

    Time::time_point start = Time::now();
    for (int value : values) {
        heap.push(value);
    }
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// 4. Heapify random
// -----------------------------

HeapifyPrepared prepare_case_heapify_random(int n) {
    return {make_random_unique_values(n, RANDOM_SEED)};
}

double measure_heapify_random_binary_heap(const HeapifyPrepared& prepared) {
    const vector<int>& values = prepared.values;

    Time::time_point start = Time::now();
    BinaryHeap<int> heap(values);
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

double measure_heapify_random_std_priority_queue(const HeapifyPrepared& prepared) {
    vector<int> data = prepared.values;

    Time::time_point start = Time::now();
    MinPriorityQueue heap(greater<int>(), std::move(data));
    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// 5. Pop all after heapify
// -----------------------------

PopAllPrepared prepare_case_pop_all_after_heapify(int n) {
    vector<int> values = make_random_unique_values(n, RANDOM_SEED);

    BinaryHeap<int> base_heap(values);

    vector<int> data = values;
    MinPriorityQueue base_std_heap(greater<int>(), std::move(data));

    return {base_heap, base_std_heap};
}

double measure_pop_all_after_heapify_binary_heap(const PopAllPrepared& prepared) {
    BinaryHeap<int> heap = prepared.heap.copy();
    size_t removed_count = 0;

    Time::time_point start = Time::now();
    while (!heap.empty()) {
        heap.pop();
        ++removed_count;
    }
    Time::time_point end = Time::now();

    benchmark_sink += removed_count;
    return Seconds(end - start).count();
}

double measure_pop_all_after_heapify_std_priority_queue(const PopAllPrepared& prepared) {
    MinPriorityQueue heap = prepared.std_heap;
    size_t removed_count = 0;

    Time::time_point start = Time::now();
    while (!heap.empty()) {
        heap.pop();
        ++removed_count;
    }
    Time::time_point end = Time::now();

    benchmark_sink += removed_count;
    return Seconds(end - start).count();
}

// -----------------------------
// 6. Alternating push/pop
// -----------------------------

AlternatingPrepared prepare_case_alternating_push_pop(int n) {
    mt19937 rng(RANDOM_SEED);
    uniform_real_distribution<double> probability(0.0, 1.0);

    vector<int> initial_values = make_random_unique_values(n / 2, RANDOM_SEED);
    vector<int> operation_values = make_random_unique_values(n, RANDOM_SEED + 1);

    vector<pair<string, int>> operations;
    operations.reserve(n);

    int current_size = static_cast<int>(initial_values.size());

    for (int value : operation_values) {
        if (current_size == 0) {
            operations.emplace_back("push", value);
            ++current_size;
            continue;
        }

        if (probability(rng) < 0.5) {
            operations.emplace_back("push", value);
            ++current_size;
        } else {
            operations.emplace_back("pop", value);
            --current_size;
        }
    }

    BinaryHeap<int> base_heap(initial_values);

    vector<int> data = initial_values;
    MinPriorityQueue base_std_heap(greater<int>(), std::move(data));

    return {operations, base_heap, base_std_heap};
}

double measure_alternating_push_pop_binary_heap(const AlternatingPrepared& prepared) {
    const vector<pair<string, int>>& operations = prepared.operations;
    BinaryHeap<int> heap = prepared.heap.copy();
    size_t operation_count = 0;

    Time::time_point start = Time::now();

    for (const auto& [operation, value] : operations) {
        if (operation == "push") {
            heap.push(value);
        } else {
            heap.pop();
        }
        ++operation_count;
    }

    Time::time_point end = Time::now();

    benchmark_sink += operation_count + heap.size();
    return Seconds(end - start).count();
}

double measure_alternating_push_pop_std_priority_queue(const AlternatingPrepared& prepared) {
    const vector<pair<string, int>>& operations = prepared.operations;
    MinPriorityQueue heap = prepared.std_heap;
    size_t operation_count = 0;

    Time::time_point start = Time::now();

    for (const auto& [operation, value] : operations) {
        if (operation == "push") {
            heap.push(value);
        } else {
            heap.pop();
        }
        ++operation_count;
    }

    Time::time_point end = Time::now();

    benchmark_sink += operation_count + heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// 7. Top-k largest
// -----------------------------

TopKPrepared prepare_case_top_k_largest(int n) {
    int k = min(TOP_K, n);
    return {make_random_unique_values(n, RANDOM_SEED), k};
}

double measure_top_k_largest_binary_heap(const TopKPrepared& prepared) {
    const vector<int>& values = prepared.values;
    int k = prepared.k;
    BinaryHeap<int> heap;

    Time::time_point start = Time::now();

    for (int value : values) {
        if (static_cast<int>(heap.size()) < k) {
            heap.push(value);
        } else if (value > heap.peek()) {
            heap.pop();
            heap.push(value);
        }
    }

    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

double measure_top_k_largest_std_priority_queue(const TopKPrepared& prepared) {
    const vector<int>& values = prepared.values;
    int k = prepared.k;
    MinPriorityQueue heap;

    Time::time_point start = Time::now();

    for (int value : values) {
        if (static_cast<int>(heap.size()) < k) {
            heap.push(value);
        } else if (value > heap.top()) {
            heap.pop();
            heap.push(value);
        }
    }

    Time::time_point end = Time::now();

    benchmark_sink += heap.size();
    return Seconds(end - start).count();
}

// -----------------------------
// Запуск всех benchmark-ов
// -----------------------------

void run_all() {
    write_csv_header(cout);

    benchmark_case<BasicPrepared>(
        cout,
        "binary_heap",
        "1. Push sequential",
        SIZES,
        prepare_case_push_sequential,
        {
            {"BinaryHeap", measure_push_sequential_binary_heap},
            {"priority_queue", measure_push_sequential_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<BasicPrepared>(
        cout,
        "binary_heap",
        "2. Push reverse sequential",
        SIZES,
        prepare_case_push_reverse_sequential,
        {
            {"BinaryHeap", measure_push_reverse_sequential_binary_heap},
            {"priority_queue", measure_push_reverse_sequential_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<BasicPrepared>(
        cout,
        "binary_heap",
        "3. Push random",
        SIZES,
        prepare_case_push_random,
        {
            {"BinaryHeap", measure_push_random_binary_heap},
            {"priority_queue", measure_push_random_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<HeapifyPrepared>(
        cout,
        "binary_heap",
        "4. Heapify random",
        SIZES,
        prepare_case_heapify_random,
        {
            {"BinaryHeap", measure_heapify_random_binary_heap},
            {"priority_queue", measure_heapify_random_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<PopAllPrepared>(
        cout,
        "binary_heap",
        "5. Pop all after heapify",
        SIZES,
        prepare_case_pop_all_after_heapify,
        {
            {"BinaryHeap", measure_pop_all_after_heapify_binary_heap},
            {"priority_queue", measure_pop_all_after_heapify_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<AlternatingPrepared>(
        cout,
        "binary_heap",
        "6. Alternating push/pop",
        SIZES,
        prepare_case_alternating_push_pop,
        {
            {"BinaryHeap", measure_alternating_push_pop_binary_heap},
            {"priority_queue", measure_alternating_push_pop_std_priority_queue}
        },
        REPEATS
    );

    benchmark_case<TopKPrepared>(
        cout,
        "binary_heap",
        "7. Top-k largest",
        SIZES,
        prepare_case_top_k_largest,
        {
            {"BinaryHeap", measure_top_k_largest_binary_heap},
            {"priority_queue", measure_top_k_largest_std_priority_queue}
        },
        REPEATS
    );
}

int main() {
    run_all();
    return 0;
}